tls/s2n_client_hello_request.c

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <s2n.h>
+
+#include "tls/s2n_connection.h"
+#include "utils/s2n_safety.h"
+
+int s2n_client_hello_request_recv(struct s2n_connection *conn)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE(conn->actual_protocol_version < S2N_TLS13, S2N_ERR_BAD_MESSAGE);
```

Comment:
Why close the connection when TLS 1.3 was negotiated, instead of ignoring the ClientHelloRequest like how it's treated in TLS 1.2?

Reply:
TLS1.2 explicitly allows the HelloRequest message to appear at any point in the connection. TLS1.3 doesn't, and a HelloRequest would be very unexpected. We could allow it anyway, but without a valid use case I'd rather we treat it as the invalid TLS1.3 message it is.